cds/utils.bal

```diff
@@ -0,0 +1,192 @@
+            return request;
+        }
+    } else {
+        return request;
+    }
+
+    if (request.prefetch is map<r4:DomainResource>) {
+        prefetchData = <map<r4:DomainResource>>request.prefetch;
+    }
+
+    // If there are no any prefetch data in the request, or
+    // required prefetch template keys are not matching with keys in the request prefetch data 
+    //then following should we given in the request to fetch the required FHIR data
+    string[] templateKeys = prefetchTemplates.keys().sort();
+    string[] dataKeys = prefetchData.keys().sort();
+    if (templateKeys != dataKeys || templateKeys.count() > dataKeys.count()) {
+        // 1. Fhir server url
+        if (request.fhirServer !is string) {
+            return createCdsError("Can not find fhirServer url in the request", 400);
+        }
+
+        // 2. FhirAuthorization object contains access token, scopes etc
+        if (request.fhirAuthorization !is FhirAuthorization) {
+            return createCdsError("Can not find fhirAuthorization in the request", 400);
+        }
+
+        // Fetch missing prefetch FHIR data
+        foreach string templateId in prefetchTemplates.keys() {
+            if (!prefetchData.hasKey(templateId)) {
+                request = check fetchFhirResource(templateId, prefetchTemplates.get(templateId), request);
+            } else {
+                r4:FHIRValidationError? validated = validator:validate(prefetchData.get(templateId));
+                if validated is r4:FHIRValidationError {
+                    return createCdsError(string `FHIR data provided for prefetch ${templateId} in the request is not valid`, 412, cause = validated);
+                }
+            }
+        }
+    } else {
+        foreach var key in prefetchData.keys() {
+            r4:FHIRValidationError? validated = validator:validate(prefetchData.get(key));
+            if validated is r4:FHIRValidationError {
+                return createCdsError(string `FHIR data provided for the prefetch: ${key}`, 400, cause = validated);
+            }
+        }
+    }
+
+    return request;
+}
+
+isolated function fetchFhirResource(string prefetchTemplateKey, string prefetchTemplate, CdsRequest cdsRequest) returns CdsError|CdsRequest {
+    string template = prefetchTemplate.clone();
+    string split = regexp:split(re `\.`, regexp:split(re `\}\}`, regexp:split(re `\{\{`, prefetchTemplate)[1])[0])[1];
+    string fhirServer = <string>cdsRequest.fhirServer;
+    anydata contextValue = cdsRequest.context.get(split);
+    if contextValue is string {
+        template = regexp:replace(re `\{\{context.${split}\}\}`, template, contextValue);
+    }
+
+    if (fhirServer.endsWith("/")) {
+        fhirServer = fhirServer.substring(0, fhirServer.length() - 1);
+    }
+
+    if (!(template.startsWith("/"))) {
+        template = string `/${template}`;
+    }
+
+    http:Client|http:ClientError fhirClient = new (fhirServer);
```

Comment:
Need to use fhir client

Reply:
We discussed this earlier: Our FHIR client is not capable of passing the headers, in this scenario, we need to pass the access token in the header.

Reply:
Is there a issue created to track this?

Reply:
Git Issue: https://github.com/wso2-enterprise/open-healthcare/issues/1620